code/modules/mob/living/carbon/human/species_types/mothmen.dm

```diff
@@ -54,3 +62,132 @@
+		return
+	if(!HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+		to_chat(H, "<span class='warning'>Your wings are fine as they are!</span>")
+		return
+	if(H.nutrition < COCOON_NUTRITION_AMOUNT)
+		to_chat(H, "<span class='warning'>You are too hungry to weave a cocoon!</span>")
+		return
+	H.visible_message("<span class='notice'>[H] begins to hold still and concentrate on weaving a cocoon...</span>", \
+	"<span class='notice'>You begin to focus on weaving a cocoon... (This will take [DisplayTimeText(COCOON_WEAVE_DELAY)] and you must hold still.)</span>")
+	H.adjustStaminaLoss(20, 0) //this is here to deter people from spamming it if they get interrupted
+	if(do_after(H, COCOON_WEAVE_DELAY, FALSE, H))
+		if(!ismoth(H))
+			to_chat(H, "<span class='warning'>You have lost your mandibles and cannot weave anymore!.</span>")
+			return
+		if(H.incapacitated())
+			to_chat(H, "<span class='warning'>You cannot weave a cocoon in your current state.</span>")
+			return
+		if(!HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+			to_chat(H, "<span class='warning'>Your wings are fine as they are!</span>")
+			return
+		H.visible_message("<span class='notice'>[H] finishes weaving a cocoon!</span>", "<span class='notice'>You finish weaving your cocoon.</span>")
+		var/obj/structure/moth_cocoon/C = new(get_turf(H))
+		H.forceMove(C)
+		H.Sleeping(20, 0)
+		C.preparing_to_emerge = TRUE
+		H.apply_status_effect(STATUS_EFFECT_COCOONED)
+		H.log_message("has finished weaving a cocoon.", LOG_GAME)
+		addtimer(CALLBACK(src, .proc/emerge, C), COCOON_EMERGE_DELAY, TIMER_UNIQUE)
+	else
+		to_chat(H, "<span class='warning'>You need to hold still in order to weave a cocoon!</span>")
+
+/datum/action/innate/cocoon/IsAvailable()
+	if(..())
+		var/mob/living/carbon/human/H = owner
+		if(HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+			return TRUE
+		return FALSE
+
+//Removes moth from cocoon, restores burnt wings
+/datum/action/innate/cocoon/proc/emerge(obj/structure/moth_cocoon/C)
+	for(var/mob/living/carbon/human/H in C.contents)
+		if(!H.has_status_effect(STATUS_EFFECT_COCOONED))
+			return
+		SEND_SIGNAL(H, COMSIG_CLEAR_MOOD_EVENT, "burnt_wings")
+		if(ismoth(H) && HAS_TRAIT(H, TRAIT_MOTH_BURNT))
+			REMOVE_TRAIT(H, TRAIT_MOTH_BURNT, "fire")
+			var/obj/item/organ/wings/moth/W = H.getorgan(/obj/item/organ/wings/moth)
+			if(W)
+				W.flight_level = WINGS_FLIGHTLESS//The check for wings getting burned makes them cosmetic, so this allows the burned off effect to be applied again
+		H.dna.species.handle_mutant_bodyparts(H)
+		H.dna.species.handle_body(H)
+	C.preparing_to_emerge = FALSE
+	qdel(C)
+
+/obj/structure/moth_cocoon
+	name = "\improper Mothperson cocoon"
+	desc = "Someone wrapped in a Mothperson cocoon. It's best to let them rest."
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "cocoon_large1"
+	anchored = TRUE
+	max_integrity = 10
+	var/preparing_to_emerge = TRUE//Determines whether or not the mothperson is done regenerating their wings with FALSE meaning they are
```

Comment:
This variable name is really unintuitive. make it like `done_regenerating` or something